libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2033,18 +2033,29 @@ void TypeChecker::typeCheckBytesConcatFunction(
 	typeCheckFunctionGeneralChecks(_functionCall, _functionType);
 
 	for (shared_ptr<Expression const> const& argument: _functionCall.arguments())
-		if (
-			Type const* argumentType = type(*argument);
+	{
+		Type const* argumentType = type(*argument);
+		bool notConvertibleToBytes = argumentType &&
 			!argumentType->isImplicitlyConvertibleTo(*TypeProvider::fixedBytes(32)) &&
-			!argumentType->isImplicitlyConvertibleTo(*TypeProvider::bytesMemory())
-		)
+			!argumentType->isImplicitlyConvertibleTo(*TypeProvider::bytesMemory());
+
+		// Only literals with all 32 bytes specified are allowed. For others the number of bytes is
```

Comment:
Can we still use `"123"` and `0x124567` and `hex"aabbcc"`? (not sure if we should for all of them)

Also would the following condition also work:
 - if it is a rational number type, it has to be implicitly convertible to bytes32 but not to bytes31

Reply:
- `"123"` and hex"aabbcc" - yes. All string literals are supported because they're implicitly convertible to `bytes` or fixed bytes.
- `0x124567` - no, because it's only convertible to `bytes4` and only literals convertible to `bytes32` are accepted.

> Also would the following condition also work:

Yes. It would accept only 32-byte-long hex literals and exclude the one with 32 zeros.

Reply:
> Yes. It would accept only 32-byte-long hex literals and exclude the one with 32 zeros.

Wait. No, it would not work because I think we do not want to exclude `0x0000000000000000000000000000000000000000000000000000000000000000`, only `0`. We need a condition that would differentiate between the two forms of zero and `isImplicitlyConvertibleTo()` behaves the same way for both.

Reply:
I added all of your examples to test cases.